File: backend/verify_platforms.py
"""Explicit integration check against the isolated RSI deployment, not arbitrary production data."""
import asyncio
import json
import os
import httpx
from .config import ARTIFACTS
from .connectors import platform_tools
from .domain import now
from .graph_store import write_private
from .runtime import create_run
from .tools import ToolContext


async def verify(source):
    base = os.getenv(source.upper() + '_BASE_URL')
    assert base == 'http://127.0.0.1:' + ('18080' if source == 'erpnext' else '18081'), 'Expected isolated experiment tunnel'
    tools = {t.name: t for t in platform_tools(source)}
    context = ToolContext(create_run({'scenario': 'finance' if source == 'erpnext' else 'support', 'mode': 'live', 'source': source, 'task': 'Platform validation'}))
    async def call(name, **args):
        return await tools[name].execute(args, context)
    rows = []
    for page in range(1, 6):
        data = await call('erpnext_list_invoices' if source == 'erpnext' else 'zammad_list_tickets', **({'status': 'all'} if source == 'erpnext' else {}), page=page, pageSize=2)
        rows.extend(data['records'])
        if not data['mayHaveMore']:
            break
    assert len(rows) == len({r.get('id', r.get('name')) for r in rows}) == 6
    if source == 'erpnext':
        assert all(r['name'].startswith('RSI-INV-') and r['currency'] == 'CNY' for r in rows)
        assert sum(r['grand_total'] for r in rows) == 66500
        assert sum(r['outstanding_amount'] for r in rows) == 24500
        payments = (await call('erpnext_list_payments', page=1, pageSize=50))['records']
        assert len(payments) == 6
        assert sum(p['reference_no'] == 'RSI-BANK-1004' for p in payments) == 2
        payment = await call('erpnext_get_payment', paymentId='RSI-PAY-002')
        assert payment['references'][0]['reference_name'] == 'RSI-INV-002' and payment['references'][0]['allocated_amount'] == 5000
        invoice = await call('erpnext_get_invoice', invoiceId='RSI-INV-002')
        assert invoice['outstanding_amount'] == 3500
        customer = await call('erpnext_get_customer', customerId=invoice['customer'])
        headers = {'Authorization': f"token {os.environ['ERPNEXT_API_KEY']}:{os.environ['ERPNEXT_API_SECRET']}"}
        path, body = '/api/resource/Customer/' + customer['name'], {'customer_name': customer['customer_name']}
        details = {'invoices': 6, 'payments': 6, 'outstandingCny': 24500, 'partialReferencesVerified': True}
    else:
        assert all(r['title'].startswith('[RSI-') for r in rows)
        states = await call('zammad_list_states')
        await call('zammad_list_priorities')
        closed = next(s['id'] for s in states if s['name'] == 'closed')
        assert sum(r['state_id'] != closed for r in rows) == 5
        ticket = await call('zammad_get_ticket', ticketId=rows[0]['id'])
        assert ticket['first_response_escalation_at']
        assert await call('zammad_get_ticket_articles', ticketId=ticket['id'])
        headers = {'Authorization': 'Token token=' + os.environ['ZAMMAD_API_TOKEN']}
        path, body = '/api/v1/tickets/' + str(ticket['id']), {'title': ticket['title']}
        async with httpx.AsyncClient(trust_env=False) as client:
            assert (await client.get(base + '/api/v1/tickets/1', headers=headers)).status_code == 403
        details = {'tickets': 6, 'activeTickets': 5, 'slaVerified': True, 'defaultGroupDenied': True}
    async with httpx.AsyncClient(trust_env=False) as client:
        response = await client.put(base + path, headers=headers, json=body)
        assert response.status_code == 403, 'Reader must not be able to update platform state'
    return {'platform': source, 'status': 'passed', 'writeStatus': 403, 'pageSize': 2, 'evidenceReferences': len(context.evidence), **details}


async def main():
    results = await asyncio.gather(verify('erpnext'), verify('zammad'))
    report = {'verifiedAt': now(), 'backend': 'python', 'results': results}
    write_private(ARTIFACTS / 'platform-deployment' / 'python-verification.json', report)
    print(json.dumps(report, ensure_ascii=False, indent=2))


if __name__ == '__main__':
    asyncio.run(main())

File: deploy/verify-platforms.ts
import '../server/config.js';
import assert from 'node:assert/strict';
import { mkdir, writeFile } from 'node:fs/promises';
import { platformTools } from '../server/connectors.js';
import { createRun } from '../server/runtime.js';
import type { ToolContext } from '../server/tools.js';

/** Tests only the explicit RSI deployment seed; never points at arbitrary production data. */
async function verify(platform: 'erpnext' | 'zammad') {
  const baseUrl = process.env[platform.toUpperCase() + '_BASE_URL'];
  assert.equal(baseUrl, `http://127.0.0.1:${platform === 'erpnext' ? 18080 : 18081}`, 'Expected the local experiment tunnel');
  const tools = platformTools(platform);
  const context: ToolContext = { run: createRun({ scenario: platform === 'erpnext' ? 'finance' : 'support', source: platform, mode: 'live', task: 'Deployment validation' }, null), signal: AbortSignal.timeout(60000), evidence: new Set() };
  async function call(name: string, args: unknown = {}) { return await tools.find(tool => tool.name === name)!.execute(args, context) as any; }
  const records: any[] = [];
  for (let page = 1; page <= 5; page++) {
    const response = await call(platform === 'erpnext' ? 'erpnext_list_invoices' : 'zammad_list_tickets', { ...(platform === 'erpnext' ? { status: 'all' } : {}), page, pageSize: 2 });
    records.push(...response.records);
    if (!response.mayHaveMore) break;
  }
  assert.equal(records.length, 6);
  assert.equal(new Set(records.map(item => item.name ?? item.id)).size, 6, 'Pagination must not duplicate records');
  let details: Record<string, unknown>;
  let writeStatus: number;
  if (platform === 'erpnext') {
    assert.ok(records.every(row => row.name.startsWith('RSI-INV-') && row.currency === 'CNY'));
    assert.equal(records.reduce((sum, row) => sum + row.grand_total, 0), 66500);
    assert.equal(records.reduce((sum, row) => sum + row.outstanding_amount, 0), 24500);
    const payments = (await call('erpnext_list_payments', { page: 1, pageSize: 50 })).records;
    assert.equal(payments.length, 6);
    assert.equal(payments.filter((row: any) => row.reference_no === 'RSI-BANK-1004').length, 2);
    const partial = await call('erpnext_get_payment', { paymentId: 'RSI-PAY-002' });
    assert.equal(partial.references[0].reference_name, 'RSI-INV-002');
    assert.equal(partial.references[0].allocated_amount, 5000);
    const invoice = await call('erpnext_get_invoice', { invoiceId: 'RSI-INV-002' });
    assert.equal(invoice.outstanding_amount, 3500);
    const customer = await call('erpnext_get_customer', { customerId: invoice.customer });
    assert.equal(customer.name, 'RSI-CUST-02');
    const headers = { Authorization: `token ${process.env.ERPNEXT_API_KEY}:${process.env.ERPNEXT_API_SECRET}`, 'Content-Type': 'application/json' };
    // A same-value update can have no business effect even if unexpectedly permitted.
    const response = await fetch(`${baseUrl}/api/resource/Customer/${customer.name}`, { method: 'PUT', headers, body: JSON.stringify({ customer_name: customer.customer_name }), signal: context.signal });
    writeStatus = response.status;
    details = { invoices: 6, payments: 6, totalCny: 66500, outstandingCny: 24500, duplicateBankReferenceCount: 2, partialPaymentReferencesVerified: true };
  } else {
    assert.ok(records.every(row => row.title.startsWith('[RSI-')));
    const states = await call('zammad_list_states');
    const priorities = await call('zammad_list_priorities');
    const closedId = states.find((row: any) => row.name === 'closed')?.id;
    assert.equal(records.filter(row => row.state_id !== closedId).length, 5);
    assert.ok(priorities.some((row: any) => row.name === '3 high'));
    const overdue = records.find(row => row.title.startsWith('[RSI-1001]'));
    assert.ok(overdue.first_response_escalation_at, 'Seed SLA must calculate an actual deadline');
    const ticket = await call('zammad_get_ticket', { ticketId: overdue.id });
    const articles = await call('zammad_get_ticket_articles', { ticketId: overdue.id });
    assert.ok(articles.length > 0); assert.equal(ticket.id, overdue.id);
    const headers = { Authorization: `Token token=${process.env.ZAMMAD_API_TOKEN}`, 'Content-Type': 'application/json' };
    const response = await fetch(`${baseUrl}/api/v1/tickets/${ticket.id}`, { method: 'PUT', headers, body: JSON.stringify({ title: ticket.title }), signal: context.signal });
    writeStatus = response.status;
    const hidden = await fetch(`${baseUrl}/api/v1/tickets/1`, { headers, signal: context.signal });
    assert.equal(hidden.status, 403, 'Default example group must remain invisible to the RSI reader');
    details = { tickets: 6, activeTickets: 5, articleReadVerified: true, slaDeadlineVerified: true, defaultGroupAccess: 'denied' };
  }
  assert.equal(writeStatus, 403, 'Platform must reject writes from the reader credentials');
  return { platform, status: 'passed', paginationPageSize: 2, writeStatus, evidenceReferences: context.evidence.size, ...details };
}

const results = await Promise.all([verify('erpnext'), verify('zammad')]);
const report = { verifiedAt: new Date().toISOString(), results };
await mkdir('artifacts/platform-deployment', { recursive: true, mode: 0o700 });
await writeFile('artifacts/platform-deployment/verification.json', JSON.stringify(report, null, 2), { mode: 0o600 });
console.log(JSON.stringify(report, null, 2));
